### PR TITLE
Support "go to definition" for SelfParams

### LINKED ISSUE #6657: 'go to definition' does not work on `self` parameter
I would expect it to go to the definition of the type that this `impl` is for. Instead it says `No definition found for 'self'`.

Specifically this happened for me on https://github.com/rust-lang/rust/blob/774bce7f5e9d0cc99a30cf37be5d175a3b530beb/compiler/rustc_trait_selection/src/traits/fulfill.rs#L311, but I expect it would happen on any function taking `self`.

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #6657

### CODE REVIEW

File: crates/ide/src/goto_definition.rs [hunk 1]
```diff
@@ -44,6 +39,12 @@ pub(crate) fn goto_definition(
                 let nav = def.try_to_nav(sema.db)?;
                 vec![nav]
             },
+            ast::SelfParam(self_param) => {
+                let ty = sema.type_of_self(&self_param)?;
+                let adt_def = ty.autoderef(db).filter_map(|ty| ty.as_adt()).last()?;
```
Comment (jonas-schievink) [inline crates/ide/src/goto_definition.rs:44]: This sounds like it won't work for the `impl u8` etc. in libcore, but that's alright

**Review by jonas-schievink — COMMENTED** [review]
bors r+

**Comment by flodiebold** [comment]
Hmm, I'm not sure I agree with that behavior. The `self` parameter *is* a definition, and go to definition on all instances of `self` in the function should go to it. There's go to type definition for going to the type, which already works. (I intended to mention that on the issue, but forgot about it...)

**Comment by Veykril** [comment]
Hm, that is a good point actually. I didn't think `go to type definition` worked here(though I didn't test it either) because we don't show a hover popup for the `self` name like we do for other names in the function signature.

**Comment by Veykril** [comment]
So I guess reverting this, and showing the hover type message like we do for other names can be considered a better "fix" for the issue?

**Comment by flodiebold** [comment]
Well, we should also make go to definition work on `self`, it should just go to the parameter. (I think VS Code will then trigger find references if you use go to definition on the parameter itself? We should also make *that* work...)

**Comment by jonas-schievink** [comment]
Yeah, I'd be fine with that behavior as well, but I've often wanted to hit F12 on the definition of `self` and expected that to take me to the type.

**Comment by jyn514** [comment]
If there's a way for `self` to go to the function signature when it's used, and to to the type in the function signature, that would be ideal. But if it's only one or the other I think I do agree that it's better to always use the function signature.

**Pull request merged**